Spark job fails when CSV datastore defines templated server definition

I'm running a spark job where my conf.xml defines a CSV datastore that defines a filename using the `{org.datacleaner.hadoop.environment}` template server. Somehow this still does not seem to work:
```
6/03/10 08:30:01 ERROR ApplicationMaster: User class threw exception: java.lang.IllegalArgumentException: Illegal character in authority at index 7: hdfs://{org.datacleaner.hadoop.environment}/datacleaner/data/companies_i18n.csv
java.lang.IllegalArgumentException: Illegal character in authority at index 7: hdfs://{org.datacleaner.hadoop.environment}/datacleaner/data/companies_i18n.csv
	at java.net.URI.create(URI.java:852)
	at org.datacleaner.spark.SparkConfigurationReaderInterceptor.createResource(SparkConfigurationReaderInterceptor.java:62)
	at org.datacleaner.configuration.JaxbConfigurationReader.createDatastore(JaxbConfigurationReader.java:1251)
	at org.datacleaner.configuration.JaxbConfigurationReader.createDatastoreCatalog(JaxbConfigurationReader.java:690)
	at org.datacleaner.configuration.JaxbConfigurationReader.create(JaxbConfigurationReader.java:286)
	at org.datacleaner.configuration.JaxbConfigurationReader.create(JaxbConfigurationReader.java:206)
	at org.datacleaner.configuration.JaxbConfigurationReader.read(JaxbConfigurationReader.java:177)
	at org.datacleaner.spark.SparkJobContext.getConfiguration(SparkJobContext.java:124)
	at org.datacleaner.spark.SparkJobContext.getAnalysisJobBuilder(SparkJobContext.java:151)
	at org.datacleaner.spark.SparkJobContext.getAnalysisJob(SparkJobContext.java:143)
	at org.datacleaner.spark.SparkAnalysisRunner.run(SparkAnalysisRunner.java:87)
	at org.datacleaner.spark.Main.main(Main.java:76)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$2.run(ApplicationMaster.scala:525)
Caused by: java.net.URISyntaxException: Illegal character in authority at index 7: hdfs://{org.datacleaner.hadoop.environment}/datacleaner/data/companies_i18n.csv
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.parseAuthority(URI.java:3186)
	at java.net.URI$Parser.parseHierarchical(URI.java:3097)
	at java.net.URI$Parser.parse(URI.java:3053)
	at java.net.URI.<init>(URI.java:588)
	at java.net.URI.create(URI.java:850)
```

--- a/engine/core/src/main/java/org/datacleaner/util/convert/HadoopResourceBuilder.java
+++ b/engine/core/src/main/java/org/datacleaner/util/convert/HadoopResourceBuilder.java
@@ -33,10 +33,22 @@ public class HadoopResourceBuilder {
     private final String _clusterReferenceName;
     private final Configuration _configuration;
 
-    private final Pattern _pattern = Pattern.compile("(?:[\\w\\+\\-\\.]+://)?\\{([\\w\\.]*)\\}(.*)");
+    /**
+     * A regular expression {@link Pattern} that matches resource URIs
+     * containing template items for the server definition, for instance:
+     * 
+     * hdfs://{myserver}/foo/bar.txt
+     * 
+     * <ul>
+     * <li>Group 1: The scheme (example 'hdfs')</li>
+     * <li>Group 2: The template name (example 'myserver')</li>
+     * <li>Group 3: The path (example '/foo/bar.txt')</li>
+     * </ul>
+     */
+    public static final Pattern RESOURCE_SCHEME_PATTERN = Pattern.compile("([\\w\\+\\-\\.]+)://\\{([\\w\\.]*)\\}(.*)");
 
     public HadoopResourceBuilder(ServerInformationCatalog catalog, String templatedUri) {
-        final Matcher matcher = _pattern.matcher(templatedUri);
+        final Matcher matcher = RESOURCE_SCHEME_PATTERN.matcher(templatedUri);
         if (!matcher.matches()) {
             _clusterReferenceName = null;
             final String fixedUri = templatedUri.replace(" ", "%20");
@@ -51,11 +63,11 @@ public class HadoopResourceBuilder {
             _configuration.set("fs.defaultFS", fixedUri);
             _uri = URI.create(fixedUri);
         } else {
-            _clusterReferenceName = matcher.group(1);
+            _clusterReferenceName = matcher.group(2);
             final HadoopClusterInformation hadoopClusterInformation = (HadoopClusterInformation) catalog.getServer(
                     _clusterReferenceName);
             _configuration = hadoopClusterInformation.getConfiguration();
-            _uri = URI.create(matcher.group(2).replace(" ", "%20"));
+            _uri = URI.create(matcher.group(3).replace(" ", "%20"));
         }
     }
 

--- a/engine/core/src/main/java/org/datacleaner/util/convert/HadoopResourceBuilder.java
+++ b/engine/core/src/main/java/org/datacleaner/util/convert/HadoopResourceBuilder.java
@@ -33,7 +33,7 @@ public class HadoopResourceBuilder {
     private final String _clusterReferenceName;
     private final Configuration _configuration;
 
-    Pattern _pattern = Pattern.compile("(?:[\\w\\+\\-\\.]+://)?\\{([\\w\\.]*)\\}(.*)");
+    private final Pattern _pattern = Pattern.compile("(?:[\\w\\+\\-\\.]+://)?\\{([\\w\\.]*)\\}(.*)");
 
     public HadoopResourceBuilder(ServerInformationCatalog catalog, String templatedUri) {
         final Matcher matcher = _pattern.matcher(templatedUri);

--- a/engine/core/src/test/java/org/datacleaner/util/convert/HadoopResourceBuilderTest.java
+++ b/engine/core/src/test/java/org/datacleaner/util/convert/HadoopResourceBuilderTest.java
@@ -1,0 +1,51 @@
+/**
+ * DataCleaner (community edition)
+ * Copyright (C) 2014 Neopost - Customer Information Management
+ *
+ * This copyrighted material is made available to anyone wishing to use, modify,
+ * copy, or redistribute it subject to the terms and conditions of the GNU
+ * Lesser General Public License, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution; if not, write to:
+ * Free Software Foundation, Inc.
+ * 51 Franklin Street, Fifth Floor
+ * Boston, MA  02110-1301  USA
+ */
+package org.datacleaner.util.convert;
+
+import static org.junit.Assert.*;
+
+import java.util.regex.Matcher;
+
+import org.junit.Test;
+
+public class HadoopResourceBuilderTest {
+
+    @Test
+    public void testPatternGroupsVanilla() throws Exception {
+        final Matcher matcher = HadoopResourceBuilder.RESOURCE_SCHEME_PATTERN.matcher("hdfs://{myserver}/foo/bar.txt");
+        assertTrue(matcher.find());
+        assertEquals("hdfs", matcher.group(1));
+        assertEquals("myserver", matcher.group(2));
+        assertEquals("/foo/bar.txt", matcher.group(3));
+    }
+
+    @Test
+    public void testPatternGroupsNoScheme() throws Exception {
+        final Matcher matcher = HadoopResourceBuilder.RESOURCE_SCHEME_PATTERN.matcher("/foo/bar.txt");
+        assertFalse(matcher.find());
+    }
+
+    @Test
+    public void testPatternGroupsNoServer() throws Exception {
+        final Matcher matcher = HadoopResourceBuilder.RESOURCE_SCHEME_PATTERN.matcher("hdfs:///foo/bar.txt");
+        assertFalse(matcher.find());
+    }
+
+}

--- a/engine/env/spark/src/main/java/org/datacleaner/spark/SparkConfigurationReaderInterceptor.java
+++ b/engine/env/spark/src/main/java/org/datacleaner/spark/SparkConfigurationReaderInterceptor.java
@@ -63,7 +63,7 @@ public class SparkConfigurationReaderInterceptor extends DefaultConfigurationRea
     public Resource createResource(String resourceUrl, DataCleanerConfiguration tempConfiguration) {
         final Matcher matcher = HadoopResourceBuilder.RESOURCE_SCHEME_PATTERN.matcher(resourceUrl);
         if (matcher.find()) {
-            resourceUrl = matcher.group(1) + ":///" + matcher.group(3);
+            resourceUrl = matcher.group(1) + "://" + matcher.group(3);
         }
         final URI uri = URI.create(resourceUrl);
         return _hdfsHelper.getResourceToUse(uri);

--- a/engine/env/spark/src/main/java/org/datacleaner/spark/SparkConfigurationReaderInterceptor.java
+++ b/engine/env/spark/src/main/java/org/datacleaner/spark/SparkConfigurationReaderInterceptor.java
@@ -21,6 +21,8 @@ package org.datacleaner.spark;
 
 import java.net.URI;
 import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 import org.apache.metamodel.util.Resource;
 import org.datacleaner.configuration.ConfigurationReaderInterceptor;
@@ -45,6 +47,9 @@ public class SparkConfigurationReaderInterceptor extends DefaultConfigurationRea
     private static final DescriptorProvider DESCRIPTOR_PROVIDER = new ClasspathScanDescriptorProvider(TASK_RUNNER)
             .scanPackage("org.datacleaner", true).scanPackage("com.hi", true).scanPackage("com.neopost", true);
     private static final StorageProvider STORAGE_PROVIDER = new InMemoryStorageProvider(500, 20);
+    
+    // used to strip e.g. {org.datacleaner.hadoop.environment} out of HDFS URIs
+    private static final Pattern STRIP_SERVER_PATTERN = Pattern.compile("(.+)://\\{.+\\}/(.+)");
 
     private static final DataCleanerEnvironment BASE_ENVIRONMENT = new DataCleanerEnvironmentImpl()
             .withTaskRunner(TASK_RUNNER).withDescriptorProvider(DESCRIPTOR_PROVIDER)
@@ -59,6 +64,10 @@ public class SparkConfigurationReaderInterceptor extends DefaultConfigurationRea
 
     @Override
     public Resource createResource(String resourceUrl, DataCleanerConfiguration tempConfiguration) {
+        final Matcher matcher = STRIP_SERVER_PATTERN.matcher(resourceUrl);
+        if (matcher.find()) {
+            resourceUrl = matcher.group(1) + ":///" + matcher.group(2);
+        }
         final URI uri = URI.create(resourceUrl);
         return _hdfsHelper.getResourceToUse(uri);
     }

--- a/engine/env/spark/src/main/java/org/datacleaner/spark/SparkConfigurationReaderInterceptor.java
+++ b/engine/env/spark/src/main/java/org/datacleaner/spark/SparkConfigurationReaderInterceptor.java
@@ -22,7 +22,6 @@ package org.datacleaner.spark;
 import java.net.URI;
 import java.util.Map;
 import java.util.regex.Matcher;
-import java.util.regex.Pattern;
 
 import org.apache.metamodel.util.Resource;
 import org.datacleaner.configuration.ConfigurationReaderInterceptor;
@@ -37,6 +36,7 @@ import org.datacleaner.job.concurrent.TaskRunner;
 import org.datacleaner.spark.utils.HdfsHelper;
 import org.datacleaner.storage.InMemoryStorageProvider;
 import org.datacleaner.storage.StorageProvider;
+import org.datacleaner.util.convert.HadoopResourceBuilder;
 
 /**
  * {@link ConfigurationReaderInterceptor} for conf.xml in a Spark environment
@@ -48,9 +48,6 @@ public class SparkConfigurationReaderInterceptor extends DefaultConfigurationRea
             .scanPackage("org.datacleaner", true).scanPackage("com.hi", true).scanPackage("com.neopost", true);
     private static final StorageProvider STORAGE_PROVIDER = new InMemoryStorageProvider(500, 20);
     
-    // used to strip e.g. {org.datacleaner.hadoop.environment} out of HDFS URIs
-    private static final Pattern STRIP_SERVER_PATTERN = Pattern.compile("(.+)://\\{.+\\}/(.+)");
-
     private static final DataCleanerEnvironment BASE_ENVIRONMENT = new DataCleanerEnvironmentImpl()
             .withTaskRunner(TASK_RUNNER).withDescriptorProvider(DESCRIPTOR_PROVIDER)
             .withStorageProvider(STORAGE_PROVIDER);
@@ -64,9 +61,9 @@ public class SparkConfigurationReaderInterceptor extends DefaultConfigurationRea
 
     @Override
     public Resource createResource(String resourceUrl, DataCleanerConfiguration tempConfiguration) {
-        final Matcher matcher = STRIP_SERVER_PATTERN.matcher(resourceUrl);
+        final Matcher matcher = HadoopResourceBuilder.RESOURCE_SCHEME_PATTERN.matcher(resourceUrl);
         if (matcher.find()) {
-            resourceUrl = matcher.group(1) + ":///" + matcher.group(2);
+            resourceUrl = matcher.group(1) + ":///" + matcher.group(3);
         }
         final URI uri = URI.create(resourceUrl);
         return _hdfsHelper.getResourceToUse(uri);

--- a/engine/env/spark/src/main/java/org/datacleaner/spark/SparkJobContext.java
+++ b/engine/env/spark/src/main/java/org/datacleaner/spark/SparkJobContext.java
@@ -32,6 +32,7 @@ import java.util.Map;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import org.apache.metamodel.util.CollectionUtils;
+import org.apache.metamodel.util.HdfsResource;
 import org.apache.spark.api.java.JavaSparkContext;
 import org.datacleaner.configuration.DataCleanerConfiguration;
 import org.datacleaner.configuration.JaxbConfigurationReader;
@@ -146,7 +147,7 @@ public class SparkJobContext implements Serializable {
     public AnalysisJobBuilder getAnalysisJobBuilder() {
         if (_analysisJobBuilder == null) {
             // set HDFS as default scheme to avoid file resources
-            SystemProperties.setIfNotSpecified(SystemProperties.DEFAULT_RESOURCE_SCHEME, "hdfs");
+            SystemProperties.setIfNotSpecified(SystemProperties.DEFAULT_RESOURCE_SCHEME, HdfsResource.SCHEME_HDFS);
 
             final DataCleanerConfiguration configuration = getConfiguration();
             final JaxbJobReader jobReader = new JaxbJobReader(configuration);

--- a/engine/env/spark/src/test/java/org/datacleaner/spark/SparkConfigurationReaderInterceptorTest.java
+++ b/engine/env/spark/src/test/java/org/datacleaner/spark/SparkConfigurationReaderInterceptorTest.java
@@ -1,0 +1,77 @@
+/**
+ * DataCleaner (community edition)
+ * Copyright (C) 2014 Neopost - Customer Information Management
+ *
+ * This copyrighted material is made available to anyone wishing to use, modify,
+ * copy, or redistribute it subject to the terms and conditions of the GNU
+ * Lesser General Public License, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution; if not, write to:
+ * Free Software Foundation, Inc.
+ * 51 Franklin Street, Fifth Floor
+ * Boston, MA  02110-1301  USA
+ */
+package org.datacleaner.spark;
+
+import static org.junit.Assert.*;
+
+import java.util.HashMap;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.metamodel.util.Resource;
+import org.datacleaner.configuration.DataCleanerConfigurationImpl;
+import org.datacleaner.spark.utils.HdfsHelper;
+import org.datacleaner.util.HadoopResource;
+import org.junit.Before;
+import org.junit.Test;
+
+public class SparkConfigurationReaderInterceptorTest {
+
+    @Before
+    public void setup() {
+        // setup code needed to register the "last known" hadoop configuration
+        // in HdfsHelper.
+        new HdfsHelper(new Configuration());
+    }
+
+    @Test
+    public void testRemoveServerFromHdfsUrl() throws Exception {
+
+        final SparkConfigurationReaderInterceptor interceptor = new SparkConfigurationReaderInterceptor(
+                new HashMap<>());
+        final Resource resource = interceptor.createResource(
+                "hdfs://{foo.bar.hadoop.environment}/datacleaner/data/companies_1.csv",
+                new DataCleanerConfigurationImpl());
+
+        assertTrue("Found: " + resource.getClass().getName(), resource instanceof HadoopResource);
+        assertEquals("hdfs:///datacleaner/data/companies_1.csv", resource.getQualifiedPath());
+    }
+
+    @Test
+    public void testSimpleHdfsUrl() throws Exception {
+        final SparkConfigurationReaderInterceptor interceptor = new SparkConfigurationReaderInterceptor(
+                new HashMap<>());
+        final Resource resource = interceptor.createResource("hdfs:///datacleaner/data/companies_2.csv",
+                new DataCleanerConfigurationImpl());
+
+        assertTrue("Found: " + resource.getClass().getName(), resource instanceof HadoopResource);
+        assertEquals("hdfs:///datacleaner/data/companies_2.csv", resource.getQualifiedPath());
+    }
+
+    @Test
+    public void testSimpleSchemeLessUrl() throws Exception {
+        final SparkConfigurationReaderInterceptor interceptor = new SparkConfigurationReaderInterceptor(
+                new HashMap<>());
+        final Resource resource = interceptor.createResource("/datacleaner/data/companies_3.csv",
+                new DataCleanerConfigurationImpl());
+
+        assertTrue("Found: " + resource.getClass().getName(), resource instanceof HadoopResource);
+        assertEquals("hdfs:///datacleaner/data/companies_3.csv", resource.getQualifiedPath());
+    }
+}

